checker-rs/src/notebook_client.rs

```diff
@@ -0,0 +1,351 @@
+            conn.write_all(login_str.as_bytes()).await?;
+            conn.flush().await?;
+
+            debug!("Waiting for login");
+            read_until_slice(conn, PROMPT, &mut response_buf).await
+        }
+        .await
+        .map_err(|e: std::io::Error| {
+            warn!("Connection-Error on Login: {}", e);
+            CheckerErrorUnfilled::from(e).with_message("Login failed")
+        })?;
+
+        if response_buf != b"Successfully logged in!\n> " {
+            info!("Unexpected response: {}", bytes_debug_repr(&response_buf));
+            return Err(CheckerError::Mumble("Login failed"));
+        }
+
+        Ok(())
+    }
+
+    #[instrument("SET_NOTE")]
+    pub async fn set_note(&mut self, note: &str) -> CheckerResult<()> {
+        info!("Deploying Note {}", note);
+
+        let conn = &mut self.conn;
+
+        let mut response_buf = Vec::with_capacity(64);
+        async {
+            let request = format!("set {}\n", note);
+            debug!("running: '{}'", request);
+            conn.write_all(request.as_bytes()).await?;
+            conn.flush().await?;
+
+            read_until_slice(conn, PROMPT, &mut response_buf).await
+        }
+        .await
+        .map_err(|e| {
+            warn!("Connection-Error: {}", e);
+            CheckerErrorUnfilled::from(e).with_message("Set-Note connection error")
+        })?;
+
+        let response = String::from_utf8(response_buf)?;
+        let note_id = response
+            .split("Note saved! ID is ")
+            .nth(1)
+            .and_then(|substr| substr.split('!').next())
+            .ok_or(CheckerError::Mumble("Failed to set note"))?;
+
+        self.user.as_mut().unwrap().note = Some(note.to_string());
+        self.user.as_mut().unwrap().note_id = Some(note_id.to_string());
+
+        Ok(())
+    }
+
+    #[instrument("GET_NOTE")]
+    pub async fn get_note(&mut self) -> CheckerResult<String> {
+        info!(
+            "Retirieving Note for user {:?}",
+            self.user.as_ref().expect("Getting note of NONE-User")
+        );
+
+        let conn = &mut self.conn;
+        let note_id = self
+            .user
+            .as_ref()
+            .unwrap()
+            .note_id
+            .as_ref()
+            .expect("No NoteId for saved user");
+
+        let mut response_buf = Vec::with_capacity(256);
+        async {
+            let request = format!("get {}\n", note_id);
+            debug!("request: '{}'", request);
+            conn.write_all(request.as_bytes()).await?;
+            conn.flush().await?;
+
+            read_until_slice(conn, PROMPT, &mut response_buf).await
+        }
+        .await
+        .map_err(|e| {
+            info!("Socketerror when retrieving note: {:?}", e);
+            CheckerErrorUnfilled::from(e).with_message("Connection error upon fetching note")
+        })?;
+
+        response_buf.truncate(response_buf.len() - 3);
+        Ok(String::from_utf8(response_buf)?)
```

Comment:
Will lead to `InternalError` on illegal utf8 (what did you create that safe utf8-function for?)

Reply:
No it doesn't, the safe utf8-function is mainly for debug-printing. It's basically like pythons representation of a bytes object, and thats what it's used for.